src/shipit_uplift/shipit_uplift/coverage_by_changeset_impl.py

```diff
@@ -0,0 +1,196 @@
+                        },
+                    },
+                    {
+                        'eq': {
+                            'repo.changeset.id12': changeset[:12],
+                        },
+                    },
+                ],
+            },
+        }))
+
+        f = r.json()['source']['file']
+        uncovered = f['uncovered']
+        covered = [e['line'] for e in f['covered']]
+
+        all_data = [0] * max(uncovered, covered)
+        for c in covered:
+            all_data[c] = 1
+
+        return all_data
+    else:
+        assert False
+
+def has_coverage(changeset):
+    if COVERAGE_SERVICE == CoverageService.COVERALLS:
+        r = requests.get('https://coveralls.io/builds/%s.json' % get_github_commit(changeset))
+        return r.status_code == requests.codes.ok
+    elif COVERAGE_SERVICE == CoverageService.CODECOV:
+        r = requests.get('https://codecov.io/api/gh/marco-c/gecko-dev/commit/%s' % get_github_commit(changeset))
+        return r.status_code == requests.codes.ok
+    elif COVERAGE_SERVICE == CoverageService.ACTIVEDATA:
+        r = requests.post('https://activedata.allizom.org/query', data=json.dumps({
+            'from': 'coverage-summary',
+            'where': {
+                'and': [
+                    {
+                        'eq': {
+                            'repo.changeset.id12': changeset[:12],
+                        }
+                    }
+                ]
+            },
+            'select': {
+                'aggregate': 'count'
+            }
+        }))
+        return r.json()['data']['count'] > 0
+    else:
+        assert False
+
+
+def get_coverage_builds(changeset):
+    r = requests.get('https://hg.mozilla.org/mozilla-central/json-rev/' + changeset)
+    push_id = int(r.json()['pushid'])
+
+    r = requests.get('https://hg.mozilla.org/mozilla-central/json-pushes?startID=%s&endID=%s' % (push_id - 8, push_id + 7))
+    print('https://hg.mozilla.org/mozilla-central/json-pushes?startID=%s&endID=%s' % (push_id - 8, push_id + 7))
+    data = r.json()
+
+    before_changeset = None
+    after_changeset = None
+    pushes = data.items()
+    pushes_before = [(pushid, pushdata) for pushid, pushdata in pushes if int(pushid) < push_id]
+    pushes_after = [(pushid, pushdata) for pushid, pushdata in pushes if int(pushid) >= push_id]
+
+    for pushid, pushdata in sorted(pushes_before, reverse=True):
+        changeset = pushdata['changesets'][-1]
+        if has_coverage(changeset):
+            before_changeset = changeset
+            break
+
+    assert before_changeset is not None
+
+    for pushid, pushdata in sorted(pushes_after):
+        changeset = pushdata['changesets'][-1]
+        if has_coverage(changeset):
+            after_changeset = changeset
+            break
+
+    assert after_changeset is not None
+
+    return (before_changeset, after_changeset)
+
+
+def generate(changeset, builds=None):
```

Comment:
docstring since it is a big function.

I would probably split some parts of this function into smaller auxiliary functions.